src/Maps/MapData.cpp

```diff
@@ -1,6 +1,51 @@
 #include "MapData.h"
 #include "CellType.h"
 #include <algorithm>
+#include <cmath>
+#include <limits>
+#include <stdexcept>
+
+MapData::MapData() :
+	versionTag(minMapVersion),
+	isSavedGame(false),
+	mapTileWidth(0),
+	mapTileHeight(0) { }
+
+MapData::MapData(const MapHeader& mapHeader) :
+	versionTag(mapHeader.versionTag),
+	isSavedGame(mapHeader.bSavedGame),
+	mapTileWidth(mapHeader.MapTileWidth()),
+	mapTileHeight(mapHeader.mapTileHeight) { }
+
+MapHeader MapData::CreateHeader() const
+{
+	MapHeader mapHeader;
+
+	mapHeader.versionTag = versionTag;
+	mapHeader.bSavedGame = isSavedGame;
+	mapHeader.lgMapTileWidth = GetWidthInTilesLog2(mapTileWidth);
+	mapHeader.mapTileHeight = mapTileHeight;
+
+	if (tilesetSources.size() > std::numeric_limits<decltype(MapHeader::tilesetCount)>::max()) {
+		throw std::runtime_error("Too many tilesets contained in map");
+	}
+
+	mapHeader.tilesetCount = static_cast<uint32_t>(tilesetSources.size());
+
+	return mapHeader;
+}
+
+uint32_t MapData::GetWidthInTilesLog2(uint32_t widthInTiles) const
+{
+	auto lgWidthInTiles = std::log2(widthInTiles);
+
+	double integerPart;
+	if (std::modf(lgWidthInTiles, &integerPart) != 0.0) {
```

Comment:
This can be done without using floating point (which as a rule should be avoided when possible). There's also an issue in that due to rounding errors, exact equality (or inequality) comparisons should generally be avoided with floating point. You need some careful analysis to know if an equality comparison is guaranteed to work.

There are some Hacker's Delight articles on finding the base 2 log:
https://graphics.stanford.edu/~seander/bithacks.html#IntegerLogObvious

Related, and only partial, you could modify just the `if` check by using:
https://graphics.stanford.edu/~seander/bithacks.html#DetermineIfPowerOf2
You won't need it if you use the above, though it's pretty cool, so worth taking a look at. :grin: